### ISSUE TITLE
Ensure compliance with more Ruff rules

### ISSUE THREAD

**Issue by eerovaher** [astropy/astropy#14818]
## What is the problem this feature will solve?

[Ruff](https://docs.astral.sh/ruff/) is a Python linter that helps to improve and maintain code quality by checking that the code complies with a customizable subset of its [lint rules](https://docs.astral.sh/ruff/rules/). Some of the Ruff rules `astropy` has chosen not to apply at all, but many are currently being ignored simply because no-one has fixed the violations in the `astropy` code base.

## Describe the desired outcome

[.ruff.toml](https://github.com/astropy/astropy/blob/main/.ruff.toml) in `astropy` is meant to be a temporary file. All rules listed in that file should eventually be either addressed by editing the `astropy` source code or permanently moved to [pyproject.toml](https://github.com/astropy/astropy/blob/main/pyproject.toml) (see #14736). At the moment many of the ignored rules are violated in only a handful of places in the code base, so they can be addressed without too much effort.

### Selecting a rule to address

To see all the temporarily ignored Ruff rule violations in `astropy` run
```shell
ruff check --config=pyproject.toml --statistics .
```
This will produce a list of rules that `astropy` currently does not comply with ordered by the number of violations. It is possible to limit the number of violations Ruff reports to a single sub-package. For example, to see the number of violations in `astropy.time` run
```shell
ruff check --config=pyproject.toml --statistics astropy/time
```

Use `ruff rule` to find out more about a rule. For example,
```shell
$ ruff rule ISC001
# single-line-implicit-string-concatenation (ISC001)

Derived from the **flake8-implicit-str-concat** linter.

## What it does
Checks for implicitly concatenated strings on a single line.

## Why is this bad?
While it is valid Python syntax to concatenate multiple string or byte
literals implicitly (via whitespace delimiters), it is unnecessary and
negatively affects code readability.

In some cases, the implicit concatenation may also be unintentional, as
autoformatters are capable of introducing single-line implicit
concatenations when collapsing long lines.

## Example
z = "The quick " "brown fox."

Use instead:
z = "The quick brown fox."
```
If a rule is not in the ignore list in [.ruff.toml](https://github.com/astropy/astropy/blob/main/.ruff.toml) and is instead in the ignore list in [pyproject.toml](https://github.com/astropy/astropy/blob/main/pyproject.toml) then it should not be addressed, but including `--config=pyproject.toml` in the `ruff check` commands should prevent any such rules from appearing in the output.

### Enforcing the rule

Different `astropy` sub-packages have different maintainers, so to simplify the review process it is preferable to address the rules one sub-package per pull request.

#### Editing Ruff configuration

Ruff must be told to enforce the rule you choose to address. In [.ruff.toml](https://github.com/astropy/astropy/blob/main/.ruff.toml) there is a global ignore list and a separate ignore list for each sub-package.
- If the  listed in its ignore list then remove it from there. 
- If the rule is in the global ignore list then remove it from there and add it to the ignore lists of sub-packages that are violating that rule, but not to the list of the sub-package you will be updating.

#### Fixing rule violations

Ruff can fix some of the violations automatically, e.g.
```shell
ruff check --config=pyproject.toml --select=C417 --fix .
```
It is highly recommended that you [set up `pre-commit`](https://docs.astropy.org/en/stable/development/workflow/development_workflow.html#pre-commit). If you do then you can fix both any automatically fixable Ruff rule violations and also any code style violations that automatic edits by Ruff might cause by invoking Ruff through the [`pre-commit` framework](https://pre-commit.com/) with
```shell
pre-commit run ruff --all
```
If Ruff cannot apply automatic fixes then the code needs to be edited manually, but Ruff can still be used to find the violations.

## Suggestions for first-time contributors

The following list does not claim to be complete and might be expanded in the future.
- [ ] [B007](https://docs.astral.sh/ruff/rules/unused-loop-control-variable/)
- [x] [C408](https://docs.astral.sh/ruff/rules/unnecessary-collection-call/) -- #14841, #15662, #15664, #17791,  #17798
- [ ] [EXE002](https://docs.astral.sh/ruff/rules/shebang-missing-executable-file/)
- [x] [PLE0101](https://docs.astral.sh/ruff/rules/return-in-init/)  #17814
- [x] [PT022](https://docs.astral.sh/ruff/rules/pytest-useless-yield-fixture/) #17479
- [ ] [RET504](https://docs.astral.sh/ruff/rules/unnecessary-assign/)

**Comment by pllim** [comment]
I thought there were disagreements at Astropy Coordination Meeting 2023 whether this qualifies as "good first issue" or not, because someone unfamiliar with the code won't know if the "corrections" actually introduce bugs?

**Comment by eerovaher** [comment]
Addressing some of the rules can be very complicated and requires familiarity with the package, but many rules are uncontroversial and very suitable for first time contributors. Eventually the simple rules will be addressed and the `Package-novice` and `good first issue` labels should be removed, but we are not close to that at the moment.

**Comment by pllim** [comment]
If you want the "novice" and "good first issue" labels back, please list in your original post clearly which ones are the uncontroversial rules. Thanks!

**Comment by Twinotters** [comment]
Hi! Hello. I'd love to participate and take this on as a first issue, unless someone is already working on it. The only issue is that I might need some guidance! Thank you in advance.

**Comment by nstarman** [comment]
Hi @Twinotters, we're always very happy for contributors and you can definitely take on `ruff` improvements as a first issue. @eerovaher has been kind enough to assemble a list of suggestions for first-time contributors, so if you see something you're interested in, we'd love a PR!

**Comment by arthurxvtv** [comment]
Hello! I have interest on working on the suggested issues. That being said, i want to know more about how do i approach them. Do i work on them together in the same PR or do i make separately PRs for each issue that i take on? If the latter, do i have to create the respective separate issues here on GitHub before making the PRs?

**Comment by nstarman** [comment]
Hi @arthurxvtv! 

> I have interest on working on the suggested issues.

🎉 

> separately PRs for each issue that i take on?

That's usually preferred. Keep the PRs atomic.

> do i have to create the respective separate issues here on GitHub before making the PRs?

I think if you hover over the TaskList that @eerovaher made you can easily convert an item to an Issue. That being said, the TaskList is fine if you don't want to open an issue before the PR.

<img width="618" alt="Screenshot 2023-05-17 at 16 42 07" src="https://github.com/astropy/astropy/assets/8949649/414ffcc4-4673-4283-b4dc-6665930e9214">

**Comment by nstarman** [comment]
@Twinotters @arthurxvtv, just collecting data. As potential contributors, do you think Pinning this issue to the top of the Issues page is useful for discovery of good first issues?

**Comment by arthurxvtv** [comment]
@pllim I think very it's useful, it can be hard sometimes to discover these issues and the pin could help a lot.

**Comment by Mubin17** [comment]
| Rules   | # of Violations |
|---------|------------|
| C400    | 3          |
| C406    | 7          |
| C408    | 219        |
| C413    | 2          |
| C416    | 39         |
| C414    | 2          |
| C417    | 1          |
| ISC001  | 18         |
| ISC003  | 17         |
| PIE790  | 41         |
| PIE804  | 1          |
| PIE810  | 1          |
| PLW0120 | 11         |
| RET504  | 116        |
| RET505  | 545        |
| RET506  | 138        |

@arthurxvtv I have created a table that compares different rules with the number of violations in the Astropy code base. Based on these numbers you can tackle all issues with less than 20 violations (fetch the rules with less than 20 violations from the above here) in a single PR.

**Comment by pllim** [comment]
Re: Issue pin -- I usually only do it if a known issue is critical (e.g., something is horribly broken). If we start pinning issues that we think are good issues, it will be crowded to the point that pinned issues are useless. I suggest you find another way. Thank you for your understanding!

**Comment by pllim** [comment]
Whoever interested to work on part of this issue, there is no need to open a new issue for each of the small parts. Just open PR and mention this issue. Thanks!

**Comment by dhomeier** [comment]
> `astropy` has chosen to apply all Ruff rules unless they are explicitly ignored

Is that in summary the outcome of the [coordination meeting breakout](https://docs.google.com/document/d/1sBImbw1p4mM3VYWGQJDRbj2N3uqP5Ed_Wfb8BDjJj2o/edit#heading=h.xtrdasd8kkom)? It would be good to have this phrased as some form of official astropy policy, as the process is possibly not entirely clear to people who have not been in the breakout.
Also if this indeed means adopting a fluid set of rules with every new ruff release, to have some more info on what governance body is deciding at ruff what rules are introduced or added.

**Comment by eerovaher** [comment]
@dhomeier, if Ruff adds a new rule that `astropy` does not comply with then it gets added to the ignore list, so updates to Ruff do not automatically lead to code changes in `astropy`.

**Comment by dhomeier** [comment]
> @dhomeier, if Ruff adds a new rule that `astropy` does not comply with then it gets added to the ignore list, so updates to Ruff do not automatically lead to code changes in `astropy`.

But the `astropy` policy as manifest in this issue is that every rule implemented by Ruff is to be removed from the ignore list at some point, by a PR linked to this issue? Or was the policy as decided in the breakout, to strive to implement all rules effective as of Ruff 0.0.263 or whatever, and any rules added since, and in the future, are to be decided on in a follow-up issue/task?

**Comment by nstarman** [comment]
The current practice is to triage when pre-commit is bumped. If it's obviously not something we want to do, then it's added to pyproject. If it's something we want to implement then it's added to the ruff.toml. There's a group of us that are "codeowners" to the pre-commit file and review the new rules.

**Comment by dhomeier** [comment]
So currently (just checked) ruff rules are indeed pinned to 0.0.263 in pre-commit. Does that mean if I have doubts about the usefulness of a certain ruff rule, I should watch (or perhaps should have in good time) updates to the pre-commit file and chime in in the discussion there? I just doubt „obviously not something we want to do“ will always necessarily be unambiguous and uncontroversial.

**Comment by nstarman** [comment]
Yes! If your interested in the ongoing discussion, add yourself as a codeowner for automatic review requests.

**Comment by dhomeier** [comment]
Thanks; stupid question: how do I make myself code owner for those files (not quite sure I really want to, but if it's the only way to keep updated what syntax rules are being/going to be imposed...)?
Had quite a bit of connection problems to GH today; maybe it's just that.

**Comment by nstarman** [comment]
It's in the file `.github/CODEOWNERS`

**Comment by Alpha-Ursae-Minoris** [comment]
While working on [my previous PR](https://github.com/astropy/astropy/pull/15033) I noticed, that ruff noted a violation of PD012 (".read_csv is preferred to .read_table; provides same functionality"). However, the only use of .read_table is in the [astropy/io/misc/parquet.py](https://github.com/astropy/astropy/blob/1d3b81abb8db3c935924b603daff2c47a6f4a9df/astropy/io/misc/parquet.py#L110) file, where it is used as pyarrow.parquet.read_table() instead of pd.read_table(). 

Based on the [parquet documentation](https://arrow.apache.org/docs/python/generated/pyarrow.parquet.read_table.html), parquet is independent of pandas and the methods simply share the same name. Parquet also has a .read_csv() method, however unlike in Pandas this method seems to not be identical to .read_table(), as .read_table() allows the input of pyarrow.NativeFile, while .read_csv() does not. 

I'm unsure whether this presents an issue within astropy/io/misc/parquet.py or not. If it does, it should probably be noted within the code. I would be happy to submit a PR to address PD012, if you could clarify this for me.

**Comment by nstarman** [comment]
> probably good reason to keep the PD rules ignored for good in pyproject.toml.

I agree. I do that in all my projects.

**Comment by pllim** [comment]
@eerovaher , can we check off C408 yet? Thanks for tracking this!

**Comment by nstarman** [comment]
@pllim C408 is only addressed in a few subpackages. In most cases it's in the per-package ignore. We should probably keep the reference open, just change the link in the checklist.

**Comment by eerovaher** [comment]
@pllim asked
> can we check off C408 yet?

```shell
$ ruff check --select C408 --config pyproject.toml --statistics .
123	C408	[*] Unnecessary `dict` call (rewrite as a literal)
```
So the answer is no.

**Comment by mhvk** [comment]
Seeing `C408` done in units in #15664, I think we should exclude this rule from being generally enforced: while it makes total sense to replace empty `tuple()` with `()`, etc., I like the ability to define `kwargs` in a way that looks just like what one does when calling a function.

p.s. For coordinates in #15662, most changes were definitely fine, though also there the last two I would have preferred not to have made.

**Comment by dhomeier** [comment]
Fully agree; I saw a lot of `dict(...)` declarations removed in io.ascii that I felt had been used by design, and for good reasons.

**Comment by nstarman** [comment]
So there are 2 good reasons why using the built-in dict constructor `{}` is better than the `dict` callable:

1. Speed. It's like 2x faster to use the dict constructor. Applying Ruff C408 is an easy speedup wherever it's fixed. 

![CleanShot 2023-12-01 at 16 36 37@2x](https://github.com/astropy/astropy/assets/8949649/8440719c-ddbd-442f-8412-e3f65d6ac525)

2. Safety. Python does not protect `dict` as a symbol so it can be re-assigned. This is relatively rare, admittedly.

![CleanShot 2023-12-01 at 16 38 21@2x](https://github.com/astropy/astropy/assets/8949649/4daebeb5-326d-49b3-9656-1f3586c36ade)

In total, `{}` is 100% equivalent but 2x faster and safer.

**Comment by eerovaher** [comment]
It seems to me that most of the instances where you would like to not enforce `C408` are in tests. We could then consider ignoring the rule in test files and adding `noqa` instructions to few specific lines outside tests. If you think this is something we should discuss then I suggest you open a separate issue. Personally I wouldn't mind enforcing `C408` in all of `astropy`.

**Comment by dhomeier** [comment]
I don't think it is simply an issue of tests vs. functional code – there are instances where readability is sacrificed for sometimes insignificant performance gains (sometimes not so, for sure) in the latter as well. Though those cases could perhaps be covered with `noqa` easily enough.

**Comment by RahatAhmed171** [comment]
Hi, I would like to contribute to fix the remaining issues. Is C408 already assigned to someone? If not, how do I get started? Thanks.

**Comment by dhomeier** [comment]
I've updated the task list to link the PRs already dealing with C408 (if I have not overlooked any that are not labelled accordingly); except for the `units` one they have already been merged, so if you check the other subpackages for violations you should be able to find the ones you can still work on.

**Comment by Rekaniry** [comment]
Hello guys! Can I try to help on this issue?

**Comment by nstarman** [comment]
Absolutely! We welcome new contributions. I would recommend reading the excellent instructions at the top of this Issue. I would recommend doing this on a per-subpackage basis. You're welcome to start jn Cosmology, where I can review any PRs.

**Comment by NathanMuniz18** [comment]
hello guys, can i try help on this?

**Comment by eerovaher** [comment]
The rules listed in the opening post are all uncontroversial. The disagreements around [C408 (unnecessary-collection-call)](https://docs.astral.sh/ruff/rules/unnecessary-collection-call) were resolved by turning on the [allow-dict-calls-with-keyword-arguments](https://docs.astral.sh/ruff/settings/#lint_flake8-comprehensions_allow-dict-calls-with-keyword-arguments) setting.

**Comment by VarunNikam22** [comment]
Is it better to fix these violations sub-package by sub-package or rule-by-rule? I started taking up `astropy/constants` but then noticed that there are also some global ignored rules which are also violated in `constants` and need to be addressed too. Should I address them separately or in the same PR?

Please advise.

**Comment by neutrinoceros** [comment]
It's best to issue fixes per sub-package, as your PRs might otherwise get blocked by just one missing reviewer. I merged your first PR quickly because I really didn't see how it could be controversial but generally it's best to target just one subpackage at a time when possible. Thanks !

**Comment by VarunNikam22** [comment]
How do you think I should I proceed with the global ignore rules then, since they would also affect other subpackages I'm not editing

**Comment by VarunNikam22** [comment]
Quick question: Are all the rules mentioned in .ruff.toml to be enforced across the package? How do we decide on ambiguous cases, if any?

For example,

- **E402** asks to shift all module imports to top of the file. Astropy violates this only in the __init__.py scripts where they need to be imported after global config variables
- **N816** discourages using `k_B` for Boltzmann constant and instead suggests something like `k_b`; but I guess that would change the way the constants are imported by users and need to be documented too

Please advise.

**Comment by neutrinoceros** [comment]
re: E402
I don't think we want to comply with this rule for `__init__.py` files. I'd argue it should be moved to permanent settings.

re: N816
This is even clearer: API stability is (by far) more important than lint compliance, so I think we should permanently ignore this rule, at least for public APIs. All existing violations in `astropy.constants`, `astropy.coordinates` and `astropy.units` should be ignored, and I think it makes sense to disable the rule completely for these subpackages. Other violations are either on private APIs or tests, so it should be fine to fix them.

**Comment by VarunNikam22** [comment]
@neutrinoceros Any reason RET505 was left in `.ruff.toml` even after making it permanent in commit cb0685fde96114988d608d8005c6b6ea582ab256 ?

**Comment by kangqiwang** [comment]
| Rules | # of Violations |
| -------- | ------- |
| B007  | 87    |
| PLE0101 | 4     |
| RET504   | 145    |

**Comment by eteq** [comment]
As discussed at Astropy 2025 coordination meeting, we think the low-hanging fruit are done, and we no longer want to encourage this as good first issue because it's diminishing returns.
